### PR TITLE
Close multiples issues on HTTP behavior

### LINKED ISSUE #316: Return message on access forbidden is not clear
`{"message": "Invalid Token: token do not have this ACL"}`

Should become 

`{"message":"Invalid Token: no permission"}%`
or
`{"message":"Invalid API key: no permission"}%`

### LINKED ISSUE #325: Don't return the `formatted` object when not format is required
We always return both documents when doing a search, the formatted and non-formatted types.
We should not return both formats if the user doesn't asks for some formatting.

### LINKED ISSUE #369: Unwrap error after updating `rankedRules`
My documents:
```json
[
    {
        "objectId": 42,
        "title": "The Hitchhiker's Guide to the Galaxy"
    },
    {
        "objectId": 4,
        "title": "Harry Potter and the Half-Blood Prince"
    },
    {
        "objectId": 123,
        "title": "Pride and Prejudice"
    },
    {
        "objectId": 1344,
        "title": "The Hobbit"
    },
    {
        "objectId": 456,
        "title": "Le Petit Prince"
    },
    {
        "objectId": 1,
        "title": "Alice In Wonderland"
    }
]
```
Steps:
- route `POST /indexes/:uid/settings` with body: `{"rankingRules": {"objectId": "asc" }}`
- search 'prince' with search route on this index.

It crashes with :
```
thread 'tokio-runtime-worker-2' panicked at 'called `Result::unwrap()` on an `Err` value: AttributeNotRegisteredForRanking', src/libcore/result.rs:1165:5
thread 'tokio-runtime-worker-3' panicked at 'called `Result::unwrap()` on an `Err` value: AttributeNotRegisteredForRanking', src/libcore/result.rs:1165:5
```

### LINKED ISSUE #372: [API] Stop-words without schema
Currently, if I have an index with no schema (because there are no documents), I can add stop-words, but I cannot delete them because: `this index does not have a schema`.

I don't know what is expected:
- Prevent from addition and deletion when there is no schema
- Allow addition and deletion even if there is no schema

### LINKED ISSUE #377: [API] Delete index route when Index Not Found
If I call the route `DELETE /indexes/:uid` and `uid` does not exist, I don't have any error message in the body (unlike all the others indexes routes)

### LINKED ISSUE #402: Reset settings
The `null` value is not taken into account by the server for the route `POST /indexes/:uid/settings`

Example:
1) Request: `POST /indexes/:uid/settings`
Body:
```json
{
     "rankingOrder": [
        "title"
    ],
    "distinctField": "",
    "rankingRules": {
        "description": "asc"
    }
}
```
2) Request: `GET /indexes/:uid/settings` 
Response:
```json
{
    "rankingOrder": [
        "title"
    ],
    "distinctField": "",
    "rankingRules": {
        "description": "asc"
    }
}
```
3) Reset with request: `POST /indexes/:uid/settings`: 
Body:
```json
{
      "rankingOrder": null,
      "distinctField": null,
      "rankingRules": null
 }
```
4) Request: `GET /indexes/:uid/settings` 
Response:
```json
{
    "rankingOrder": [
        "title"
    ],
    "distinctField": "",
    "rankingRules": {
        "description": "asc"
    }
}
```

### PR DESCRIPTION
**Pull request by qdequele**
closes #377 
A valid message is now returned if there is no index.
Capture d’écran 2019-12-12 à 14 04 45

closes #372 
A stop word addition without schema is now processed.
Capture d’écran 2019-12-12 à 15 34 02

closes #316 
The response for API Keys is now more clear.
Capture d’écran 2019-12-12 à 15 50 05

Closes #325 
Return formatted section on search only if it's requested.
Capture d’écran 2019-12-12 à 16 22 33
Capture d’écran 2019-12-12 à 16 22 24

Closes #369 
This doesn't unwrap anymore

touch #397

closes #402 
Check settings, there are all to null
Capture d’écran 2019-12-13 à 10 17 13
Push new settings
Capture d’écran 2019-12-13 à 10 17 30
Check settings, they have the right values
Capture d’écran 2019-12-13 à 10 17 35
Remove settings
Capture d’écran 2019-12-13 à 10 17 48
Check settings, there are all to null
Capture d’écran 2019-12-13 à 10 17 57

### CODE REVIEW

**Comment by Kerollmops** [comment]
Concerning the warning/error handling when serializing the user data. Things I thought about were:

  - Add a `&mut Vec<String>` [the `Indexer` struct](https://github.com/meilisearch/MeiliSearch/blob/5ad73fe08b443639d5b3019977cbf72b40f96c21/meilisearch-core/src/serde/indexer.rs#L9-L13) and push the error message in it
  - Add another field named `user_document_id` which represent the human readable document id, its type would be a String
  - Retrieved it from [the `extract_document_id` function](https://github.com/meilisearch/MeiliSearch/blob/5ad73fe08b443639d5b3019977cbf72b40f96c21/meilisearch-core/src/serde/extract_document_id.rs#L10-L19) which will return a tuple `(DocumentId, String)`.
  - We need to [keep both the `DocumentId` and the String](https://github.com/meilisearch/MeiliSearch/blob/5ad73fe08b443639d5b3019977cbf72b40f96c21/meilisearch-core/src/serde/extract_document_id.rs#L248-L254) that has been hashed to produce this id.

File: meilisearch-core/src/serde/serializer.rs [hunk 1]
```diff
@@ -320,13 +321,17 @@ where
...
+        match value.serialize(indexer) {
+            Ok(Some(number_of_words)) => {
+                documents_fields_counts.put_document_field_count(
+                    txn,
+                    document_id,
+                    attribute,
+                    number_of_words as u64,
+                )?;
+            },
+            Ok(None) => {},
```
Comment (Kerollmops) [inline meilisearch-core/src/serde/serializer.rs]: prefer using `()` instead of an empty scope.

File: meilisearch-core/src/serde/serializer.rs [hunk 2]
```diff
@@ -320,13 +321,17 @@ where
...
+            Ok(Some(number_of_words)) => {
+                documents_fields_counts.put_document_field_count(
+                    txn,
+                    document_id,
+                    attribute,
+                    number_of_words as u64,
+                )?;
+            },
+            Ok(None) => (),
+            Err(err) => warn!("Indexing error; {}", err),
```
Comment (Kerollmops) [inline meilisearch-core/src/serde/serializer.rs]: Please revert that to the previous `if let Some`.

File: meilisearch-core/src/update/stop_words_deletion.rs [hunk 3]
```diff
@@ -101,14 +101,18 @@ pub fn apply_stop_words_deletion(
...
     // lets reindex everything...
-    reindex_all_documents(
-        writer,
-        main_store,
-        documents_fields_store,
-        documents_fields_counts_store,
-        postings_lists_store,
-        docs_words_store,
-    )?;
+    if let Ok(number) = main_store.number_of_documents(writer) {
```
Comment (Kerollmops) [inline meilisearch-core/src/update/stop_words_deletion.rs]: I do not understand what would be the impact of this change.

File: meilisearch-core/src/serde/serializer.rs (hunk 2 above)
Comment (Kerollmops) [reply meilisearch-core/src/serde/serializer.rs]: By doing this you discard every potential errors even those related to write to disk in LMDB.

File: meilisearch-core/src/update/stop_words_deletion.rs (hunk 3 above)
Comment (qdequele) [reply meilisearch-core/src/update/stop_words_deletion.rs]: no documents => nothing to reindex and also no problem of missing schema

File: meilisearch-core/src/serde/serializer.rs (hunk 2 above)
Comment (qdequele) [reply meilisearch-core/src/serde/serializer.rs]: Ok, I got it

File: meilisearch-core/src/update/stop_words_deletion.rs (hunk 3 above)
Comment (Kerollmops) [reply meilisearch-core/src/update/stop_words_deletion.rs]: Ho I understand now, the first step of the `reindex_all_documents` is to gather the schema and removing stop-words triggers a reindexing. Thank you!

File: meilisearch-http/src/routes/index.rs [hunk 4]
```diff
@@ -390,19 +390,10 @@ pub async fn get_all_updates_status(ctx: Context<Data>) -> SResult<Response> {
 
 pub async fn delete_index(ctx: Context<Data>) -> SResult<StatusCode> {
     ctx.is_allowed(IndexesWrite)?;
+    let _ = ctx.index()?;
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/index.rs]: Ok, so now the 404 is raised by [the index method](https://github.com/meilisearch/MeiliSearch/blob/bd51bec397b2e9ecc7377e2fd42a53a4d0753af4/meilisearch-http/src/helpers/tide.rs#L93-L101).

**Comment by Kerollmops** [comment]
The stop-words error was raised on stop-words deletion not addition.
https://github.com/meilisearch/MeiliSearch/pull/400/commits/88b3c051551ad49e3c0ac94ea89dcc27d1d02fb8#diff-56b030e9acd5f91a213486828900546c

**Comment by Kerollmops** [comment]
That's an awesome PR here! Great job 💯

File: meilisearch-http/src/routes/setting.rs [hunk 5]
```diff
@@ -12,7 +12,7 @@ use crate::routes::document::IndexUpdateResponse;
 use crate::Data;
 
 #[derive(Default, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase", deny_unknown_fields)]
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/setting.rs]: Why did you removed this attribute?

Comment (qdequele) [reply meilisearch-http/src/routes/setting.rs]: Sorry :/ a test related to https://github.com/meilisearch/MeiliSearch/pull/400/commits/4e486d72f1bc2db737fef311cf1065abd19ba828. Doesn't take it into account.

File: meilisearch-http/src/routes/setting.rs [hunk 6]
```diff
@@ -39,12 +39,31 @@ pub async fn get(ctx: Context<Data>) -> SResult<Response> {
...
+}
+
+// Any value that is present is considered Some value, including null.
+fn deserialize_some<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
+    where T: Deserialize<'de>,
+          D: Deserializer<'de>
+{
+    Deserialize::deserialize(deserializer).map(Some)
+}
+
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/setting.rs]: What is that new struct ?

File: meilisearch-http/src/routes/setting.rs [hunk 7]
```diff
@@ -39,12 +39,31 @@ pub async fn get(ctx: Context<Data>) -> SResult<Response> {
...
     let settings = match index.main.customs(&reader).unwrap() {
         Some(bytes) => bincode::deserialize(bytes).unwrap(),
-        None => SettingBody::default(),
+        None => Setting::default(),
     };
 
     Ok(tide::response::json(settings))
 }
 
+#[derive(Default, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/setting.rs]: Why does it impl Serialize?

File: meilisearch-http/src/routes/setting.rs (hunk 6 above)
Comment (Kerollmops) [reply meilisearch-http/src/routes/setting.rs]: Does it work? Could you provide a screenshot?

Comment (Kerollmops) [reply meilisearch-http/src/routes/setting.rs]: Did you looked at `#[serde(flatten)]`?

It is probably what you want.
https://github.com/serde-rs/serde/issues/1255

Comment (qdequele) [reply meilisearch-http/src/routes/setting.rs]: Flatten is not what I want. I need to know if the user has omitted the field or if he has set it to null. I used the example from here https://github.com/serde-rs/serde/issues/984

**Comment by bidoubiwa** [comment]
No documentation needed

**Pull request merged**